kythe/go/serving/explore/explore_test.go

```diff
@@ -0,0 +1,478 @@
+			Type:    srvpb.Callgraph_CALLER,
+		},
+	}
+)
+
+func TestChildren_badData(t *testing.T) {
+	svc := Construct(t)
+
+	reply, err := svc.Children(ctx, &epb.ChildrenRequest{
+		Tickets: []string{badType},
+	})
+	if err == nil {
+		t.Errorf("Expected Children error for bad data, got: %v", reply)
+	}
+}
+
+func TestChildren_noData(t *testing.T) {
+	svc := Construct(t)
+
+	reply, err := svc.Children(ctx, &epb.ChildrenRequest{
+		Tickets: []string{"key with no data"},
+	})
+	if err != nil {
+		testutil.FatalOnErrT(t, "Children error: %v", err)
+	}
+	if len(reply.InputToChildren) != 0 {
+		t.Errorf("Expected empty response for missing key, got: %v", reply)
+	}
+}
+
+func TestChildren(t *testing.T) {
+	svc := Construct(t)
+	request := &epb.ChildrenRequest{
+		Tickets: []string{p1},
+	}
+
+	reply, err := svc.Children(ctx, request)
+	testutil.FatalOnErrT(t, "Children error: %v", err)
+
+	expectedInputToChildren := map[string]*epb.Tickets{
+		p1: {
+			Tickets: []string{p1c1, p1c2},
+		},
+	}
+
+	if !equalTicketsKeys(reply.InputToChildren, expectedInputToChildren) {
+		t.Errorf("Expected results and actual results have different ticket key sets:\n"+
+			"expected map: %v\n actual map: %v", expectedInputToChildren, reply.InputToChildren)
+	}
+
+	for ticket, expectedChildren := range expectedInputToChildren {
+		children := reply.InputToChildren[ticket]
+		if !reflect.DeepEqual(listToSet(expectedChildren.Tickets), listToSet(children.Tickets)) {
+			t.Errorf("Mismatch for children of ticket %s; expected:\n%v actual:\n%v",
+				ticket, expectedChildren, children)
+		}
+	}
+}
+
+func TestParents_badData(t *testing.T) {
+	svc := Construct(t)
+
+	reply, err := svc.Parents(ctx, &epb.ParentsRequest{
+		Tickets: []string{badType},
+	})
+	if err == nil {
+		t.Errorf("Expected Parents error for bad data, got: %v", reply)
+	}
+}
+
+func TestParents_noData(t *testing.T) {
+	svc := Construct(t)
+
+	reply, err := svc.Parents(ctx, &epb.ParentsRequest{
+		Tickets: []string{"key with no data"},
+	})
+	if err != nil {
+		testutil.FatalOnErrT(t, "Parents error: %v", err)
+	}
+	if len(reply.InputToParents) != 0 {
+		t.Errorf("Expected empty response for missing key, got: %v", reply)
+	}
+}
+
+func TestParents(t *testing.T) {
+	svc := Construct(t)
+	request := &epb.ParentsRequest{
+		Tickets: []string{p1c1},
+	}
+
+	reply, err := svc.Parents(ctx, request)
+	testutil.FatalOnErrT(t, "Parents error: %v", err)
+
+	expectedInputToParents := map[string]*epb.Tickets{
+		p1c1: {
+			Tickets: []string{p1},
+		},
+	}
+
+	if !equalTicketsKeys(reply.InputToParents, expectedInputToParents) {
+		t.Errorf("Expected results and actual results have different ticket key sets:\n"+
+			"expected map: %v\n actual map: %v", expectedInputToParents, reply.InputToParents)
+	}
+
+	for ticket, expectedParents := range expectedInputToParents {
+		parents := reply.InputToParents[ticket]
+		if !reflect.DeepEqual(listToSet(expectedParents.Tickets), listToSet(parents.Tickets)) {
```

Comment:
Should we be testing sets of the tickets?  That would eliminate cardinality differences.  I think just sorting the lists should be enough.